Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+    get {
+      // No basename means path is root or empty (which is the dirname)
+      if self.basename == nil { return self }
+      return FilePath(self.components.dropLast())
+    }
+  }
+
+  /// The extension of the file or directoy last component.
+  ///
+  /// If `basename` is `nil` or one of the special path components
+  /// `.` or `..`, `get` returns `nil` and `set` does nothing.
+  ///
+  /// If `basename` does not contain a `.` anywhere, or only
+  /// at the start, `get` returns `nil` and `set` will append a
+  /// `.` and `newValue` to `basename`
+  ///
+  /// Otherwise `get` returns everyting after it, and `set` will
+  /// replace the extension
+  ///
+  /// Examples:
+  ///   * `/tmp/foo.txt => "txt"`
+  ///   * `/Appliations/Foo.app/ => app`
+  ///   * `/Appliations/Foo.app/bar.txt => txt`
+  ///   * `/tmp/.hidden => nil``
+  ///   * `/tmp/.. => nil``
+  ///
+  /// REVIEW NOTE: Rust has a set_extension instead, which returns whether it
+  /// did anything, but having a setter is more convenient.
+  public var `extension`: String? {
+    get {
+      fatalError()
+    }
+    set { fatalError() }
+  }
+
+  /// The non-extension portion of the file or directoy last component.
+  ///
+  /// Returns `nil` if `basename` is `nil`
+  ///
+  ///   * `/tmp/foo.txt => "foo"`
+  ///   * `/Appliations/Foo.app/ => Foo`
+  ///   * `/Appliations/Foo.app/bar.txt => bar`
+  ///   * `/tmp/.hidden => .hidden``
+  ///   * `/tmp/.. => ..``
+  ///   * `/ => nil`
+  ///
+  /// TODO: setter? That's a little harder to imagine than for `extension`.
+  /// Setting to `nil` is more likely a bug.
+  public var stem: String? {
+    fatalError()
+  }
+
+}
+
+// Modification and concatenation API
+extension FilePath {
+  /// Remove the contents, keeping the null termiantor
+  ///
+  /// TODO: better name? I like `clear` but that's not "Swifty"
+  /// TODO: Since we'll still have the null terminator, should we intern the empty path?
+  public mutating func removeContents(keepingCapacity: Bool = false) {
+    fatalError()
+  }
+
+  /// Reserve enough storage space to store `minimumCapacity` path characters
+  public mutating func reserveCapacity(_ minimumCapacity: Int) {
+    self.storage.reserveCapacity(1 + minimumCapacity)
+  }
+
+  /// Append the contents of `other`.
+  ///
+  /// TODO: What if `other` is absolute? C++ does replacement for append
+  /// but not for concat... Rust's `push` does replacement. Should we just do push/pop?
+  ///
+  public mutating func append(_ other: FilePath) {
+    // TODO: Faster to do byte copy, checking for leading separator
+    self.components.append(contentsOf: other.components)
+  }
+
+  /// Remove and return the last component of this file path. If the path is
+  /// root or empty, does nothing and returns `nil`.
+  ///
+  /// Examples:
+  /// * `"/".pop() == nil          // path is /`
+  /// * `"/foo/bar".pop() == "bar" // path is /foo`
+  public mutating func pop() -> FilePath.Component? {
```

Comment:
I would suggest `popLast` here to be a bit more clear (not a big fan of the word "pop" for data structures that aren't specifically stacks, but there's certainly precedent for that in Swift).